### ISSUE TITLE
Better support for internal PHP classes default parameters

### ISSUE THREAD

**Issue by Ocramius** [Ocramius/ProxyManager#162]
As @guilhermeblanco suggested, we can fully support internal PHP classes methods' default parameter values by using a trick like following.

Assuming this is an internal php class, and that we cannot discover the default value for `$foo` via reflection:

``` php
class SomeInternalClass
{
    public function doStuff($foo = 'bar') {}
}
```

This code should become like following in the proxy:

``` php
class MyProxy extends SomeInternalClass
{
    public function doStuff($foo = ProxyManager::PLACEHOLDER)
    {
        if (ProxyManager::PLACEHOLDER === $foo) {
            parent::doStuff(); // ignore parameter, use defaults
        } else {
            parent::doStuff($foo);
        }
    }
}
```

That will allow us to fully support default values in proxies for internal PHP classes.

Please note that this does not apply to class-hinted parameters, as the default value for a class-hinted parameter can only be `null`

**Comment by Ocramius** [comment]
Removed from 2.0.0 milestone: while this can be implemented, I'd rather have something like https://github.com/Roave/BetterReflection deal with this sort of problem instead.

**Comment by ilya-levin-lokalise** [comment]
Hi guys, first of all `null is pretty much always the default` this is not true.
Second of all, it's better to specify this behavior somewhere on the front page of the README with red text, because in fact, this library doesn't fully support proxying of internal classes. Moreover, it created **buggy** proxies.
Simple example: I wanted to proxy `Redis` built-in class. Everything works well until I call `multi` method. It has the following signature in the initial class: `public function multi($mode = Redis::MULTI) {}`
And proxy-class **changes** it to: `public function multi($mode = null);`
And then I'm getting Exceptions like `Redis::multi(): Unknown mode sent to Redis::multi`. Which is basically an irrecoverable error. Which then means, this library broke my code.
So please, take it more seriously and at least put a disclaimer.

**Comment by Ocramius** [comment]
> So please, take it more seriously and at least put a disclaimer.

Sorry, but unless you have practical advice on how to resolve the issue, this is not constructive feedback.

Patches welcome, but I can't do much about internal symbols reflection being incomplete, because that's just how the engine behaves for types not defined in userland.

**Comment by Ocramius** [comment]
For your specific Redis issues, check the related issues here:

 * https://github.com/symfony/symfony/issues/23066
 * https://github.com/BedrockStreaming/RedisMock/issues/46
 * https://github.com/Ocramius/ProxyManager/issues/461
 * https://github.com/snc/SncRedisBundle/issues/442

Specifically, from that last issue:

> Btw, this is neither a bug in ProxyManager nor in this bundle: you will need to either fix this in the extension, or provide a userland shim that can be reflected.

**Comment by ilya-levin-lokalise** [comment]
Putting a disclaimer that library have edge-cases in README **is** a constructive feedback

**Comment by alekitto** [comment]
Trying to be constructive through a proposal: what about loading class definitions for these edge-cases from a stub file? Could it be viable?

**Comment by Ocramius** [comment]
> Putting a disclaimer that library have edge-cases in README **is** a constructive feedback

Please re-read your comment, then, and perhaps ask someone to analyze it for you, so you can understand why it is not an acceptable human interaction:

 * with strangers you don't know personally
 * that provide you with a free service
 * that put lots of effort in it
 * that operate very seriously, with a long and regular track record
 * that, even after your inconstructive feedback, took the time to provide you with a set of references to similar problems, including potential workarounds

**Comment by Ocramius** [comment]
> Trying to be constructive through a proposal: what about loading class definitions for these edge-cases from a stub file? Could it be viable?

@alekitto yes, see:

 * https://github.com/Ocramius/ProxyManager/issues/162#issuecomment-168055437
 * https://github.com/Ocramius/ProxyManager/issues/49

Beware that any mis-alignment between stubs and actual runtime symbols can still lead to problems.

Using something like `BetterReflection` is viable, but requires major rework of internals.

**Comment by azjezz** [comment]
> @mnapoli we can randomize/generate the value to ensure that it is always unique

I think the best approach would be to use a combination of time, current directory/filename, and a pseudo random id.

```php
$placeholder = uniqid(true, '__proxy_manager') . time() . hash('md5', __FILE__);
```

> md5 because performance

**Comment by nicolas-grekas** [comment]
There's only one sensible way fix to the issue with `Redis` or similar: send a patch to https://github.com/phpredis/phpredis/ It's an actively maintained lib. Asking OSS authors to write and maintain piles of code that are just workarounds is a waste of their time. Even if one doesn't know C, one can patch the definitions in the source code by using copy/paste and imitation. Please consider sending them a PR @ilya-levin-lokalise!

As far as this issue is concerned, what about closing it @Ocramius, for the reason I just gave? The PHP engine is getting stricter and stricter regarding type declarations, it's not like extension authors have any excuses.

**Comment by cmb69** [comment]
It seems that issue has [already been resolved](https://github.com/phpredis/phpredis/commit/7009fa02caaad97fc353fb11aab2d6c0c1995661). Proper reflection should be available as of 5.3.5RC1 in combination with PHP 8.

**Comment by Ocramius** [comment]
> As far as this issue is concerned, what about closing it @Ocramius, for the reason I just gave? The PHP engine is getting stricter and stricter regarding type declarations, it's not like extension authors have any excuses.

Depends on whether this can still occur in the engine at all 🤔

**Comment by nicolas-grekas** [comment]
> Depends on whether this can still occur in the engine at all thinking

I think yes, having reflection out of sync from actual declaration is still possible, but it's now way easier to have them in sync AFAIK.
Maybe @nikic has some more info on the topic?

**Comment by cmb69** [comment]
> Depends on whether this can still occur in the engine at all 🤔

Yes, it [still can happen](https://3v4l.org/BenkS). In some cases, a default cannot be specified (in this case because the parameter has no default; in other cases because the function may be overloaded like `array_keys`), and of course in some cases, because an extension has not yet been updated. Wrt. to the core and bundled extensions, there are [only "few" such cases](https://heap.space/search?project=PHP-8.1&full=&defs=&refs=UNKNOWN&path=&hist=&type=php).

**Comment by Ocramius** [comment]
Yeah, problem is potentially still with `ext-*` stuff, including proprietary software. For proprietary stuff, there is a hard stop in such cases, because there would be no stubs either.

**Comment by nicolas-grekas** [comment]
Idea (likely a bad one): use a variadic argument in place of this UNKNOWN default?

**Comment by azjezz** [comment]
actually that's a pretty good one, using 

```
function foo($a = null, /* signature becomes broken here, use variadic argument */ ...$rest)
```

for

```
function foo($a = null, $b = UNKNOWN, $c = null)
```

could work, people get auto-completion based on the original class, so the signature of the proxy doesn't matter.

https://3v4l.org/u0V0a

**Comment by Ocramius** [comment]
Overall potentially feasible, need to build some trickery for point-cut proxies, where the parameters are passed to interceptor callables (and keys may be missing), but it is likely feasible 🤔

**Comment by nicolas-grekas** [comment]
Another idea: set `null` as default value and forward the call with `parent::doStuff(...func_get_args())`.
